### DISCUSSION TITLE
Object Oriented API (OOP)

### DISCUSSION THREAD

**Discussion by hinell** [fastify/fastify#4816] [Q&A] [answered]
https://github.com/fastify/fastify/blob/f4a8f3624c87ab23bec79280b876ad955ac5fafa/lib/route.js#L55-L62

Why fastify doesn't stick to OOP concept by providing basic class for Route? Why don't you extend `FindMyWay` and export it? 

Other libraries like Hapi and ExpressJS provide their basic classes that can be extended and reused. 

I wish your API be more extensible. I also would like to override basic classes and provide them to Fastify so I don't use wrappers in controlling handlers:

```js
// app.js

class MyResponse extends fastify.Response { ... }
class MyRouter extends fastify.Router { ... }

fastify.context.Response = MyResponse
fastify.context.Router = MyRouter

...

```

**Comment by jsumners** [comment]
Fastify uses factory functions to build new instances by extending the core prototypes with the provided options. This is a common [OOP pattern](https://en.wikipedia.org/wiki/Factory_(object-oriented_programming)), particularly in prototypal JavaScript. It is central to how Fastify provides the speed benefits it promises.

**Reply by hinell** [reply]
Well factory assumes that class is given to it to produce instances. This is not the case as you invoke `FindMyWay` from a module scope, which is not good. 

This way it breaks ability to compose & inherit classes.

Assuming I can inherit from `FindMyWay`, how does one passes new subclass to the `buildRouting(...)` factory ?

**Reply by metcoder95** [reply]
Do you have a special use case in which `FindMayWay` needs to be altered, and it's not covered by `fastify`?
What @jsumners describes is the chore of Fastify, and the `decorators` fulfills in most cases the need to extend the core "classes" of the framework without compromising performance.

**Reply by hinell** [reply]
I don't need it be altered. I want to _extend_ it. This is a huge difference. I would like to extend router at least and provide it to the fastify as a dependency replacing built-in one.

Decorators aren't the good way to go. Too many projects trying to play-around inheritance and break dependency-injection. This is a non-starter really.

**Reply by mcollina** [reply, marked as answer]
Thanks for letting us know your opinion. I’m glad that there are plenty of other frameworks to choose from. Hapi is great!
